Random failure in do_nd test

Looks like some test may be leaking a non closed awg instrument

```
=================================== FAILURES ===================================
______________________________ test_datasaver_1d _______________________________
[gw1] linux -- Python 3.10.13 /opt/hostedtoolcache/Python/3.10.13/x64/bin/python

experiment = test-experiment#test-sample#1@/tmp/pytest-of-runner/pytest-0/popen-gw1/test_datasaver_1d0/temp.db
--------------------...1-dummy_dac_ch1,dummy_dmm_v1-31
12-results-12-dummy_dac_ch1,dummy_dmm_v1-[76](https://github.com/QCoDeS/Qcodes/actions/runs/6974993548/job/18987846779?pr=5547#step:9:77)
13-results-13-dummy_dac_ch1,dummy_dmm_v1-76
DAC = <DummyInstrument: dummy_dac>, DMM = <DummyInstrument: dummy_dmm>
caplog = <_pytest.logging.LogCaptureFixture object at 0x7f1ce0b3beb0>
n_points = 76

    @given(n_points=hst.integers(min_value=1, max_value=[100](https://github.com/QCoDeS/Qcodes/actions/runs/6974993548/job/18987846779?pr=5547#step:9:101)))
    @example(n_points=5)
    @settings(deadline=None, suppress_health_check=(HealthCheck.function_scoped_fixture,))
    def test_datasaver_1d(
        experiment, DAC, DMM, caplog: LogCaptureFixture, n_points
    ) -> None:
        meas = Measurement()
        meas.register_parameter(DAC.ch1)
        meas.register_parameter(DMM.v1, setpoints=(DAC.ch1,))
    
        n_points_expected = 5
    
        meas.set_shapes({DMM.v1.full_name: (n_points_expected,)})
    
        with meas.run() as datasaver:
    
            for set_v in np.linspace(0, 1, n_points):
                DAC.ch1()
                datasaver.add_result((DAC.ch1, set_v),
                                     (DMM.v1, DMM.v1()))
    
        ds = datasaver.dataset
        caplog.clear()
        data = ds.get_parameter_data()
    
        for dataarray in data[DMM.v1.full_name].values():
            assert dataarray.shape == (n_points,)
    
        if n_points == n_points_expected:
            assert len(caplog.record_tuples) == 0
        elif n_points > n_points_expected:
>           assert len(caplog.record_tuples) == 2
E           AssertionError: assert 4 == 2
E            +  where 4 = len([('qcodes.instrument.visa', 20, 'Closing VISA handle to awg_sim as there are no non weak references to the instrument....a shape for dummy_dac_ch1 in dataset dummy_dmm_v1 from metadata when loading but found inconsistent lengths 76 and 5')])
E            +    where [('qcodes.instrument.visa', 20, 'Closing VISA handle to awg_sim as there are no non weak references to the instrument....a shape for dummy_dac_ch1 in dataset dummy_dmm_v1 from metadata when loading but found inconsistent lengths 76 and 5')] = <_pytest.logging.LogCaptureFixture object at 0x7f1ce0b3beb0>.record_tuples

tests/dataset/measurement/test_shapes.py:42: AssertionError

The above exception was the direct cause of the following exception:
```

we've seen similar failures  in the unit tests
[full log](https://cache.nixos.org/log/xq0cp80zffciidvk2n5b3ya2ndbbyzlz-python3.11-qcodes-0.42.1.drv)
```console
=========================== short test summary info ============================
FAILED tests/drivers/test_keithley_s46.py::test_query_close_once_at_init - assert 2 == 1
FAILED tests/parameter/test_parameter_ramp.py::test_step_ramp - assert 4 == 1
```
[full log](https://logs.ofborg.org/?attempt_id=237e6cca-42be-4a1a-98f1-21529787f4eb&key=nixos%2Fnixpkgs.288404)
```console
=========================== short test summary info ============================
FAILED tests/parameter/test_parameter_ramp.py::test_step_ramp - assert 2 == 1
```

--- a/src/qcodes/instrument/visa.py
+++ b/src/qcodes/instrument/visa.py
@@ -28,12 +28,16 @@ log = logging.getLogger(__name__)
 def _close_visa_handle(
     handle: pyvisa.resources.MessageBasedResource, name: str
 ) -> None:
-    log.info(
-        "Closing VISA handle to %s as there are no non weak "
-        "references to the instrument.",
-        name,
-    )
-    handle.close()
+    if (
+        handle.session is not None
+    ):  # pyvisa sets the session of a handle to None when it is closed
+        log.info(
+            "Closing VISA handle to %s as there are no non weak "
+            "references to the instrument.",
+            name,
+        )
+
+        handle.close()
 
 class VisaInstrument(Instrument):
 

